packages/devtools_app/lib/src/framework/release_notes/release_notes.dart

```
@@ -76,66 +75,121 @@ class ReleaseNotesController extends SidePanelController {
+      _emptyAndClose();
+      return;
+    }
+
+    final Map<String, Object?> releaseIndex;
     try {
-      // Try all patch versions for this major.minor combination until we find
-      // release notes (e.g. 2.11.4 -> 2.11.3 -> 2.11.2 -> ...).
-      final attemptedVersions = <String>[];
-      var attempts = notesVersion.patch;
-      while (attempts >= 0 && notesVersion > versionFloor) {
-        final versionString = notesVersion.toString();
+      final releaseIndexUrl =
+          _flutterDocsSite.replace(path: '/f/devtools-releases.json');
+      releaseIndex =
+          jsonDecode(await http.read(releaseIndexUrl)) as Map<String, Object?>;
+    } catch (e) {
+      _emptyAndClose(e.toString());
+      return;
+    }
+
+    final releases = releaseIndex['releases'];
+    if (releases is! Map<String, Object?>) {
+      _emptyAndClose(
+        'The DevTools release index file was incorrectly formatted.',
+      );
+      return;
+    }
+
+    // If the version floor has the same major and minor version,
+    // don't check below its patch version.
+    final int minimumPatch;
+    if (versionFloor.major == notesVersion.major &&
+        versionFloor.minor == notesVersion.minor) {
+      minimumPatch = versionFloor.patch;
+    } else {
+      minimumPatch = 0;
+    }
+
+    final majorMinor = '${notesVersion.major}.${notesVersion.minor}';
+    var patchToCheck = notesVersion.patch;
+
+    // Try each patch version in this major.minor combination until we find
+    // release notes (e.g. 2.11.4 -> 2.11.3 -> 2.11.2 -> ...).
+    while (patchToCheck >= minimumPatch) {
+      final releaseToCheck = '$majorMinor.$patchToCheck';
+      if (releases[releaseToCheck] case final String releaseNotePath) {
+        final String releaseNotesMarkdown;
         try {
-          String releaseNotesMarkdown = await http.read(
-            Uri.parse(_debugReleaseNotesUrl ?? _releaseNotesUrl(versionString)),
+          releaseNotesMarkdown = await http.read(
```

will this still print errors to the command line? When I've looked at this before, I could never get the try / catch to swallow the exception that is printed to the chrome console

---

Yeah it does, but now this is handling an error case rather than part of the core discovery behavior, so it should never happen.

I'm assuming Chrome DevTools always adds that, but I'll see if I can prevent it somehow.

---

Since `releaseToCheck` has to be a key in the `releases` map, this shouldn't fail though right? This would really only catch the case where we had a broken url for the `releaseToCheck` entry on the flutter.dev lookup

---

I was thinking more about caching or strange CDN issues/delays as the broken URL would be caught by the website's CI currently. Or maybe if the user lost connection momentarily. So you're right, in 99.99% of the cases, it shouldn't fail.

I added a comment to expand on when it could somehow fail and added a log message with the version number and the link tried so it's easier to identify what caused it to fail, if someone ever reports it.